bindings/rust/src/lib.rs

```diff
@@ -2,24 +2,31 @@
 // Licensed under the Apache License, Version 2.0, see LICENSE for details.
 // SPDX-License-Identifier: Apache-2.0
 
+#![cfg_attr(not(feature = "std"), no_std)]
```

Comment:
You should define features in `Cargo.toml`. Even though it will work like that, it is not easily discoverable.

Reply:
It's a conscious choice. There is no reason to make it discoverable. If user attempts to compile on **a** bare-metal platform it's simply engaged in the build script.

Reply:
But why should `blst` be unlike any other ordinary Rust crate? There is an established pattern in the ecosystem how this things are done, what is the point of deviating from it?

Reply:
Without having this feature defined in `Cargo.toml` and set by default you essentially compile `no_std` version for ALL PLATFORMS. Users will have to set `std` feature to avoid this, but due to "conscious choice" they will have a hard time knowing it exists in the first place.

You can probably define it from within `build.rs` but you've literally fighting the best practices in the ecosystem for the sake of confusing users.

Reply:
> you essentially compile `no_std` version for ALL PLATFORMS.

No. build.rs enables `std` feature on ALL PLATFORMS **but** bare-metal ones.

> Users will have to set std feature to avoid this,

No. Users won't have to do anything. Nor would they have to do anything special when using blst as dependency on a bare-metal platform.

Reply:
I see, I should have checked beforehand.

But you need to know that platform is bare metal or not. Idiomatic way in Rust is to expose:
* `std` feature for cases when it is required for some optional features
* `alloc` feature for cases when library can work without heap allocation, but may require/benefit from it for some features

If library supports multi-threading, it must also be a separate feature that users opt-in if they want and if they want/need it to be single-threaded, then they don't specify it.

By inferring those things and adding environment variables to control some of the behavior instead you basically make it impossible to use in some cases, impossible to change behavior in others and less discoverable in the rest because anyone who knows how typical Rust crate works now has to learn how `blst` works because it is special.

Even for you as maintainer it already cost more complicated `build.rs` that needs to be maintained and there will be users with platforms (including those that have unofficial std/no_std environment) that will either have to fork the project, bother you with adding support for platforms that do not exist in official Rust repository, all because of the design decisions made here.

This is just not how things are done in Rust and there are reasons for status quo, some of which I outlined above, but I'm sure there are others.

Reply:
> But you need to know that platform is bare metal or not.

Yes, we do. Well, rather it's apparently appropriate, if not safe, to assume that platforms what have `none` for target_os are those that require `no_std`.

Reply:
What is you have `target_os`, but don't want `std`?

When someone builds a static library for x64-64 with Rust they may want to not pull a big chunk of Rust standard library into it through `blst`. They may want to have no threads either because for their use case threading makes more sense on higher level and `blst` would just cause quadratic number of threads only making things worse.

Those assumptions are precisely the kinds things that would make it impossible to use the library as is, but that would function perfectly fine if you didn't make such assumptions and just let user specify what they want. If you CAN have `std` it doesn't mean you WANT to have `std`.

Reply:
You can also have multi-threading support with custom wasm32 implementation that would use [SharedArrayBuffer](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer) and host function calls to implement threading and surrounding primitives in browser. You make that impossible too by forcing onto user what you think they want.

I think I can continue pulling more of these examples that defy your assumptions if I try hard enough. As I said, there are good reasons why idiomatic Rust looks the way it does.

Reply:
Actually, I believe `wasm32-unknown-emscripten` already supports threads, but due to the design of the library right now user can't take advantage of it because library hardcodes in `build.rs` that threads are disabled for wasm32 :man_shrugging:

Reply:
While it's true that emscripten supports threads, the real question is whether or not Rust `threadpool` can **reliably** access the interface. As soon as `threadpool` works, the decision will be reevaluated. blst is about what actually works, not what might work.

Reply:
> They may want to have no threads either because for their use case threading makes more sense on higher level

And they can do that. Nobody actually did to my knowledge, defaulting to multi-threading appears to be generally appreciated.

> and `blst` would just cause quadratic number of threads only making things worse.

This is not true. blst uses a private **fixed-sized** threadpool.

Reply:
> blst is about what actually works, not what might work.

What I'm trying to say is `blst` can't possibly know and doesn't really need to know what works and what doesn't. I have given a few examples like support for threads where you think there isn't such support and forcing `std` on users that can, but don't want to depend on it. Both of those are not possible to achieve with current design of the library.

Simply offering two features enabled by default (`std` and `threads`) allows any imaginable supported platform will work under full control of the developer using the library.

Reply:
And what I'm saying is that if we don't know something, we make no assumptions and don't really want users to make such assumptions either. [As effectively reflected in README.md.] But as already said, it's not right spot to discuss this. Why-not-this-way is off-topic in this specific context. Open an issue or choose another channel to communicate with supranational.

Reply:
I was thinking to replace manual thread pool management with rayon, which inherits thread pool from outer context and that way you can change number of threads from the outside (or not bother and use global default thread pool), but didn't have time to do the change and bench it.